### ISSUE TITLE
increase power of IDummyDefinition and IFakeConfigurator 

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#402]
About a year ago, in https://github.com/FakeItEasy/FakeItEasy/issues/156#issuecomment-28899986 @patrik-hagne said

> … it would be simpler if the Task-dummies where implemented using IDummyDefinitions, I realize that there needs to be a change in the way dummy definitions work for that though

This stuck in my mind, and it's come back to me with the recent activity in #358. There's currently no good way to create dummy definitions for generic types—we only allow definitions for specific types.  `IFakeConfigurator` has the same restriction. The latter touches on the answer that @philippdolder and I gave for Stack Overflow question [FakeItEasy sometimes fails to create a fake when tests are run in parallel](http://stackoverflow.com/questions/26501952/fakeiteasy-sometimes-fails-to-create-a-fake-when-tests-are-run-in-parallel). Were we able to define a `IFakeConfigurator` that acted on any type, it would simplify the process of synchronizing _all_ fakes. There are likely other benefits.

I propose changing the interface definitions to allow user-specified type matching. I'll use `IDummyDefinition` as an example. Instead of

```csharp
public interface IDummyDefinition
{
    Type ForType { get; }
    object CreateDummy();
}
```

we could have 

```csharp
public interface IDummyDefinition
{
    int Priority { get; }
    bool CanCreateDummyOfType(Type type);
    object CreateDummyOfType(Type type);   
}
```

`Priority` would be used to choose between matching definitions, if there's more than one, just as `IArgumentValueFormatter.Priority` is. The methods should be self-explanatory.

This would be a breaking change, but the effects could be greatly mitigated by updating `DummyDefinition<T>`. I'll send a PR that shows how it can be changed so dummy definitions extending `DummyDefinition<T>` won't have to be touched.

**Comment by blairconrad** [comment]
Any objection to releasing 2.0.0-beta003 with this on Monday?

**Comment by adamralph** [comment]
None at all! :shipit:

**Comment by patrik-hagne** [comment]
Is this pulled into an official release? I think it's all great, but I think the interface should be renamed. IDummyCreator or similar would make more sense now.

**Comment by blairconrad** [comment]
Hey, Mr. Hägne!

It's in [2.0.0-beta003](https://www.nuget.org/packages/FakeItEasy/2.0.0-beta003) right now. No official release.

We could easily rename the interface, since we're already inconveniencing people who wrote their  `IDummyDefinition`s by hand anyhow. I'd been avoiding inconveniencing those who were relying on `DummyDefinition<T>` up 'til now, though - with the change I made, they shouldn't notice anything.
Were you thinking to rename that class as well?

**Comment by adamralph** [comment]
If we're changing the name, I vote for adhering to the `...Factory` convention.

**Comment by blairconrad** [comment]
+1

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
